libsolidity/codegen/ir/IRGenerate.cpp

```diff
@@ -0,0 +1,212 @@
+#include <libdevcore/SHA3.h>
+#include <libdevcore/CommonData.h>
+
+#include <boost/algorithm/string/predicate.hpp>
+
+using namespace std;
+using namespace dev;
+using namespace dev::solidity;
+
+bool IRGenerate::visit(ContractDefinition const& _contract)
+{
+	solUnimplementedAssert(
+		_contract.contractKind() == ContractDefinition::ContractKind::Contract,
+		"Non-contracts (libraries, interfaces) are not supported yet."
+	);
+	solUnimplementedAssert(_contract.baseContracts().empty(), "Inheritance not supported yet.");
+	solUnimplementedAssert(_contract.definedStructs().empty(), "User-defined types not supported yet.");
+	solUnimplementedAssert(_contract.definedEnums().empty(), "User-defined types not supported yet.");
+	solUnimplementedAssert(_contract.events().empty(), "Events not supported yet.");
+	solUnimplementedAssert(_contract.functionModifiers().empty(), "Modifiers not supported yet.");
+
+	solAssert(m_contracts[_contract.fullyQualifiedName()] == nullptr, "");
+	m_body = make_shared<assembly::Block>();
+	m_body->location = _contract.location();
+	m_contracts[_contract.fullyQualifiedName()] = m_body;
+
+	ASTNode::listAccept(_contract.definedFunctions(), *this);
+
+	buildDispatcher(_contract);
+
+	return false;
+}
+
+void IRGenerate::buildDispatcher(ContractDefinition const& _contract)
+{
+	appendFunction(R"(
+	{
+		// Revert if value was received.
+		function ensureNoValueTransfer()
+		{
+			switch callvalue()
+			case 0:u256 {}
+			default { revert(0:u256, 0:u256) }
+		}
+
+		// Extract 32 bit method identifier
+		function extractCallSignature() -> sig:u256
+		{
+			// FIXME: replace with constant
+			sig := div(calldataload(0:u256), exp(2:u256, 224:u256))
+		}
+	}
+	)");
+
+	assembly::Switch _switch;
+	_switch.expression = make_shared<assembly::Statement>(createFunctionCall("extractCallSignature"));
+
+	for (auto const& function: _contract.definedFunctions())
+	{
+		if (!function->isPartOfExternalInterface())
+			continue;
+
+		assembly::Literal literal;
+		literal.kind = assembly::LiteralKind::Number;
+		literal.value = toHex(FixedHash<4>::Arith(FixedHash<4>(dev::keccak256(function->externalSignature()))), HexPrefix::Add);
+		literal.type = "u256";
+
+		assembly::Block body;
+		if (!function->isPayable())
+			body.statements.emplace_back(createFunctionCall("ensureNoValueTransfer"));
+		body.statements.emplace_back(createFunctionCall("_" + function->name()));
+
+		assembly::Case _case;
+		_case.value = make_shared<assembly::Literal>(literal);
+		_case.body = std::move(body);
+
+		_switch.cases.emplace_back(_case);
+	}
+
+	assembly::Case defaultCase;
+	if (auto const& fallbackFunction = _contract.fallbackFunction())
+	{
+		assembly::Block body;
+		if (!fallbackFunction->isPayable())
+			body.statements.emplace_back(createFunctionCall("ensureNoValueTransfer"));
+		body.statements.emplace_back(createFunctionCall("fallback"));
+		defaultCase.body = std::move(body);
+	}
+	else
+		defaultCase.body = wrapInBlock(createRevert());
```

Comment:
We'll once someone writes it. We kind of agreed to merge the above way tentatively so that there's at a least a skeleton API to work with.